distributed/event.py

```diff
@@ -0,0 +1,251 @@
+            future = event.wait()
+            if timeout is not None:
+                future = asyncio.wait_for(future, timeout)
+
+            self._waiter_count[name] += 1
+            try:
+                await future
+            except TimeoutError:
+                return False
+            finally:
+                self._waiter_count[name] -= 1
+
+                if not self._waiter_count[name] and not event.is_set():
+                    # No one is waiting for this
+                    # and as the default flag for an event is false
+                    # we can safely remove it
+                    self._delete_event(name)
+
+            return True
+
+    async def event_set(self, stream=None, name=None):
+        """ Set the event with the given name to true.
+
+        All waiters on this event will be notified.
+        """
+        with log_errors():
+            name = self._normalize_name(name)
+            # No matter if someone is listening or not,
+            # we set the event to true
+            event = self._get_or_create_event(name)
+            self.scheduler.loop.add_callback(event.set)
+
+    async def event_clear(self, stream=None, name=None):
+        """Set the event with the given name to false."""
+        with log_errors():
+            name = self._normalize_name(name)
+            if not self._waiter_count[name]:
+                # No one is waiting for this
+                # and as the default flag for an event is false
+                # we can safely remove it
+                self._delete_event(name)
+
+            else:
+                # There are waiters
+                # This can happen if an event is "double-cleared"
+                # In principle, the event should be unset at this point
+                # (because if it is set, all waiters should have been
+                # notified). But to prevent race conditions
+                # do to unlucky timing, we clear anyways
+                assert name in self._events
+                event = self._events[name]
+                self.scheduler.loop.add_callback(event.clear)
+
+    async def event_is_set(self, stream=None, name=None):
+        with log_errors():
+            name = self._normalize_name(name)
+            # the default flag value is false
+            if name not in self._events:
+                return False
+
+            return self._events[name].is_set()
+
+    def _normalize_name(self, name):
+        """ Helper function to normalize an event name """
+        if isinstance(name, list):
+            name = tuple(name)
+
+        return name
+
+    def _get_or_create_event(self, name):
+        """ Helper function to return or create and return an event """
+        if name not in self._events:
+            # It seems we are the first one accessing it,
+            # so lets create a new event
+            self._events[name] = asyncio.Event()
+
+        return self._events[name]
```

Comment:
I *think* that `self._events = collections.defaultdict(asyncio.Event)`  would work here, but I could be wrong.

Reply:
Yes, makes a lot of sense!